estimators/all_estimators.py

```diff
@@ -0,0 +1,279 @@
+    except KeyError:
+        mse = history.history['mse'][-1]
+
+    bias = history.history['mae'][-1]
+    logger.debug('Calculate Predictions')
+    y_pred, y_obs, t = _predict_base_models(data=data_all, model=model, steps=param_method['steps'])
+    logger.debug('Done Predictions')
+    return y_pred, mse, bias, t, y_obs
+
+
+def _predict_base_models(data, model, steps):
+    """Predicts the outcome on the full data.
+
+    :param data: tf.data.Dataset.
+    :param model: tf.keras.Model.
+    :return: arrays with predicted outcome, observed outcome, and treat. assignment.
+    """
+    y_pred = []
+    y_obs = []
+    t = []
+    for i, (batch_x, batch_y, batch_t) in enumerate(data):
+        y_pred.append(model.predict_on_batch(batch_x))
+        y_obs.append(batch_y.numpy())
+        t.append(batch_t.numpy())
+        if i > steps:
+            break
+
+    y_pred_flat = np.concatenate(y_pred).ravel()
+    y_obs_flat = np.concatenate(y_obs).ravel()
+    t_flat = np.concatenate(t).ravel()
+
+    return np.array(y_pred_flat), np.array(y_obs_flat), np.array(t_flat)
+
+
+def _calculate_propensity_score(param_method, data, y_treated_hat, y_control_hat, t, y):
+    """ Calculating Propensity Score and return the treatment effect.
+
+    :param param_method: dictionary.
+    :param data: tf.data.Dataset.
+    :param y_treated_hat: [], predicted outcome if treated.
+    :param y_control_hat: [], predicted outcome if control.
+    :param t: [], treatment assignment.
+    :param y: [], observed outcome.
+    :return: tau (int, treat effect), var (int, variance)
+    """
+    t = t.ravel()
+    y = y.ravel()
+
+    # Calculating the Propensity Score.
+    if param_method['learn_prop_score']:
+        data_all = data.dataset_all_ps
+        # Propensity score using a logistic regression.
+        prop_score = param_method['prop_score']
+        prop_score.fit(data_all)
+        e = prop_score.predict_proba(data_all)
+        # Removing extreme values
+        # print('y shaoes', y.shape, type(y))
+        g = e[:, 1].ravel()
+        y = _truncate_by_g(y.reshape(-1, 1), g, 0.005)
+        t = _truncate_by_g(t, g, 0.005)
+        y_treated_hat = _truncate_by_g(y_treated_hat, g, 0.005)
+        y_control_hat = _truncate_by_g(y_control_hat, g, 0.005)
+        e = _truncate_by_g(e, g, 0.005)
+
+        # Estimating the treatment effect.
+        y_dif = (y_treated_hat - y_control_hat)
+        sample_size = len(y)
+        residual_treated = (t * (y - y_treated_hat))
+        residual_control = ((1 - t) * (y - y_control_hat))
+
+        residual_treated = (residual_treated / e[:, 1].ravel())
+        residual_control = (residual_control / e[:, 0].ravel())
+
+    else:
+        # Estimating the treatment effect.
+        y_dif = (y_treated_hat - y_control_hat)
+        sample_size = len(y)
+        residual_treated = (t * (y - y_treated_hat))
+        residual_control = ((1 - t) * (y - y_control_hat))
+        residual_treated = (residual_treated / 0.5)
```

Comment:
Why divide by `0.5`?

Reply:
As we talked about it, during the experiments we could run the AIPW without training a model for the propensity score. In that case, the propensity score was fixed as 0.5. I made a small modification, where P(T=1|X) = sum(t)/len(t).